core/src/service/search_service.rs

```diff
@@ -49,4 +67,345 @@ impl RequestContext<'_, '_> {
+
+                for paragraph in content.split("\n\n") {
+                    if let Some(fuzzy_match) = FuzzySearch::new(search, paragraph)
+                        .case_insensitive()
+                        .score_with(&Scoring::emphasize_distance())
+                        .best_match()
+                    {
+                        let score = fuzzy_match.score() as i64;
+
+                        if score >= LOWEST_CONTENT_SCORE_THRESHOLD {
+                            let (paragraph, matched_indices) =
+                                RequestContext::optimize_searched_text(
+                                    paragraph,
+                                    fuzzy_match.matched_indices().cloned().collect(),
+                                )?;
+
+                            content_matches.push(ContentMatch {
+                                paragraph,
+                                matched_indices,
+                                score,
+                            });
+                        }
+                    }
+                }
+
+                if !content_matches.is_empty() {
+                    if *no_matches {
+                        *no_matches = false;
+                    }
+
+                    if !should_continue.load(atomic::Ordering::Relaxed) {
+                        return Ok(());
+                    }
+
+                    if results_tx
+                        .send(SearchResult::FileContentMatches {
+                            id: info.id,
+                            path: info.path.clone(),
+                            content_matches,
+                        })
+                        .is_err()
+                    {
+                        break;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn optimize_searched_text(
+        paragraph: &str, matched_indices: Vec<usize>,
+    ) -> Result<(String, Vec<usize>), UnexpectedError> {
+        if paragraph.len() <= IDEAL_CONTENT_MATCH_LENGTH {
+            return Ok((paragraph.to_string(), matched_indices));
+        }
+
+        let mut index_offset: usize = 0;
+        let mut new_paragraph = paragraph.to_string();
+        let mut new_indices = matched_indices;
+
+        let first_match = new_indices.first().ok_or_else(|| {
+            warn!("A fuzzy match happened but there are no matched indices.");
+            UnexpectedError("No matched indices.".to_string())
+        })?;
+
+        let last_match = new_indices.last().ok_or_else(|| {
+            warn!("A fuzzy match happened but there are no matched indices.");
+            UnexpectedError("No matched indices.".to_string())
+        })?;
+
+        if *last_match < IDEAL_CONTENT_MATCH_LENGTH {
+            new_paragraph = new_paragraph
+                .chars()
+                .take(IDEAL_CONTENT_MATCH_LENGTH + CONTENT_MATCH_PADDING)
+                .collect();
+
+            new_paragraph.push_str("...");
```

Comment:
The following compiles, but can you make sure it still works as expected, and then commit it if you feel like it's better?

```suggestion
                .take(IDEAL_CONTENT_MATCH_LENGTH + CONTENT_MATCH_PADDING)
                .chain("...".chars())
                .collect();
```